domain/miningmanager/miningmanager_test.go

```diff
@@ -0,0 +1,392 @@
+		if err != nil {
+			t.Fatalf("HandleNewBlockTransactions: %v", err)
+		}
+		if contains(transactionInTheMempool, miningManager.AllTransactions()) {
+			t.Fatalf("The transaction %v, shouldn't be in the mempool, since at least one output was already spent.", transactionInTheMempool)
+		}
+	})
+}
+
+// TestOrphanTransactions verifies that a transaction could be a part of a new block template, only if it's not an orphan.
+func TestOrphanTransactions(t *testing.T) {
+	testutils.ForAllNets(t, true, func(t *testing.T, params *dagconfig.Params) {
+
+		params.BlockCoinbaseMaturity = 0
+		factory := consensus.NewFactory()
+		tc, teardown, err := factory.NewTestConsensus(params, false, "TestOrphanTransactions")
+		if err != nil {
+			t.Fatalf("Error setting up TestConsensus: %+v", err)
+		}
+		defer teardown(false)
+
+		miningFactory := miningmanager.NewFactory()
+		miningManager := miningFactory.NewMiningManager(tc, blockMaxMass, false)
+		// Before each parent transaction, We will add two blocks by consensus in order to fund the parent transactions.
+		parentTransactions, childTransactions, err := createArraysOfParentAndChildrenTransactions(params, tc)
+		if err != nil {
+			t.Fatalf("Error in createArraysOfParentAndChildrenTransactions: %v", err)
+		}
+		for _, orphanTransaction := range childTransactions {
+			err = miningManager.ValidateAndInsertTransaction(orphanTransaction, true)
+			if err != nil {
+				t.Fatalf("ValidateAndInsertTransaction: %v", err)
+			}
+		}
+		transactionsMempool := miningManager.AllTransactions()
+		for _, transaction := range transactionsMempool {
+			if contains(transaction, childTransactions) {
+				t.Fatalf("Error: an orphan transaction is exist in the mempool")
+			}
+		}
+
+		block, err := miningManager.GetBlockTemplate(&externalapi.DomainCoinbaseData{
+			ScriptPublicKey: &externalapi.ScriptPublicKey{Script: nil, Version: 0},
+			ExtraData:       nil})
+		if err != nil {
+			t.Fatalf("Failed get a block template: %v", err)
+		}
+		for _, transactionFromBlock := range block.Transactions[1:] {
+			for _, orphanTransaction := range childTransactions {
+				if consensushashing.TransactionID(transactionFromBlock) == consensushashing.TransactionID(orphanTransaction) {
+					t.Fatalf("Tranasaction with unknown parents is exist in a block that was built from GetTemplate option.")
+				}
+			}
+		}
+
+		blockParentsTransactionsHash, _, err := tc.AddBlock([]*externalapi.DomainHash{params.GenesisHash}, nil, parentTransactions)
+		if err != nil {
+			t.Fatalf("AddBlock: %v", err)
+		}
+		blockParentsTransactions, err := tc.GetBlock(blockParentsTransactionsHash)
+		if err != nil {
+			t.Fatalf("GetBlock: %v", err)
+		}
+		_, err = miningManager.HandleNewBlockTransactions(blockParentsTransactions.Transactions)
+		if err != nil {
+			t.Fatalf("HandleNewBlockTransactions: %v", err)
+		}
+		transactionsMempool = miningManager.AllTransactions()
+		for _, transaction := range transactionsMempool {
```

Comment:
transactionsMempool is currently empty, so this check will always pass. You should instead check the opposite, that all the child transactions are in the mempool